include/sparrow/dynamic_bitset.hpp

```diff
@@ -0,0 +1,338 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cassert>
+#include <climits>
+
+#include "sparrow/buffer.hpp"
+
+namespace sparrow
+{
+    template <class B>
+    class dynamic_bitset_base
+    {
+    public:
+
+        using self_type = dynamic_bitset_base<B>;
+        using storage_type = B;
+        using block_type = typename storage_type::value_type;
+        using value_type = bool;
+        using size_type = typename storage_type::size_type;
+
+        size_type size() const noexcept;
+        size_type null_count() const noexcept;
+
+        bool test(size_type pos) const;
+        void set(size_type pos, value_type value);
+
+        block_type* data() noexcept;
+        const block_type* data() const noexcept;
+        size_type block_count() const noexcept;
+
+        void swap(self_type&) noexcept;
+
+    protected:
+
+        dynamic_bitset_base(storage_type&& buffer, size_type size);
+        dynamic_bitset_base(storage_type&& buffer, size_type size, size_type null_count);
+        ~dynamic_bitset_base() = default;
+
+        dynamic_bitset_base(const dynamic_bitset_base&) = default;
+        dynamic_bitset_base(dynamic_bitset_base&&) = default;
+
+        dynamic_bitset_base& operator=(const dynamic_bitset_base&) = default;
+        dynamic_bitset_base& operator=(dynamic_bitset_base&&) = default;
+
+        void resize(size_type n, value_type = false);
+
+        size_type compute_block_count(size_type bits_count) const noexcept;
+
+    private:
+
+        size_type count_non_null() const noexcept;
+        size_type block_index(size_type pos) const noexcept;
+        size_type bit_index(size_type pos) const noexcept;
+        block_type bit_mask(size_type pos) const noexcept;
+        size_type count_extra_bits() const noexcept;
+        void zero_unused_bits();
+
+        static constexpr std::size_t s_bits_per_block = sizeof(block_type) * CHAR_BIT;
+
+        storage_type m_buffer;
+        size_type m_size;
+        size_type m_null_count;
+    };
+
+    template <class T>
+    class dynamic_bitset : public dynamic_bitset_base<buffer<T>>
```

Comment:
These types needs documentation :) Also the constructors because of the differences of ownership.

Question: as-is, `dynamic_bitset<T>` is *not* a range (begin/end/empty are missing). Is this on purpose?